[5.0] [Serialization] Use full target architectures for swiftmodule files

jrose-apple opened the pull request:
- **Explanation**: Swiftmodules in frameworks are stored under the name of their "major architecture" (the name used by `#if arch(…)`), but that creates complications for platforms that have more than one architecture variant per major architecture, like armv7 and armv7s, even if the content is "the same".

- **Scope**: In practice, only affects how the compiler imports frameworks, which only Xcode builds.

- **Issue**: rdar://problem/45174692

- **Risk**: Low. This change is backwards-compatible for all shipping Apple platforms, and non-Apple platforms don't use multi-architecture Swift modules.

- **Testing**: Passed regression tests and the source compat suite.

- **Reviewed by**: @rjmccall, @jakepetroules

jrose-apple commented:
@swift-ci Please test

AnnaZaks reviewed (approved):
Approved to be merged once the tests pass.

jrose-apple committed:
[Serialization] Use full target architectures for swiftmodule files (#21053)

Previously these used the same "major architecture" names that the
`#if os(...)` query accepts, but that can be a problem when building
for, say, both armv7 and armv7s, even if the content is "the same".

For a transition period where external build tools are involved, the
compiler will look for "arm.swiftmodule" if it fails to find
"armv7.swiftmodule" or any other 32-bit ARM architecture. No other
Apple platform architectures are affected, and AFAIK no one's using
the architecture-based layout on Linux or any other platforms.

rdar://problem/45174692
(cherry picked from commit 06f3c11377169f8577e9579da8fc72c376d16ddc)

jrose-apple commented:
\*sigh\* Resolved the master/5.0 differences wrong.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Source compat failures are UPASSs; what's more, they're UPASSs in SwiftPM projects, which are not affected by this.
